File: src/main/java/com/github/thorben/webnd/application/controller/api/GameSpecificApi.java
package com.github.thorben.webnd.application.controller.api;

import com.github.thorben.webnd.application.SessionStorage;
import com.github.thorben.webnd.application.controller.api.dto.SkillTypeList;
import com.github.thorben.webnd.application.dto.CreationDto;
import com.github.thorben.webnd.domain.character.*;
import com.github.thorben.webnd.domain.games.GameType;
import com.github.thorben.webnd.domain.games.GameTypeRepository;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RestController;

import javax.inject.Inject;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Optional;

@RestController
public class GameSpecificApi {

	@Inject
	private SkillTypeRepository skillTypeRepository;
	@Inject
	private DetailTypeRepository detailTypeRepository;
	@Inject
	private AttributeTypeRepository attributeTypeRepository;
	@Inject
	private GameTypeRepository gameTypeRepository;
	@Inject
	private BaseValueTypeRepository baseValueRepository;
	@Inject
	private SessionStorage sessionStorage;

	@GetMapping(value = "/api/game_types", produces = MediaType.APPLICATION_JSON_UTF8_VALUE)
	public ResponseEntity<List<GameType>> getAllGameTypes() {
		List<GameType> result = new ArrayList<>();
		gameTypeRepository.findAll().forEach(result::add);
		return ResponseEntity.ok(result);
	}

	@GetMapping(value = "/api/game_types/{gameId}/skills", produces = MediaType.APPLICATION_JSON_UTF8_VALUE)
	public ResponseEntity<List<SkillType>> getAllSkills(@PathVariable Integer gameId) {
		Optional<GameType> byId = gameTypeRepository.findById(gameId);
		if (!byId.isPresent()) {
			return ResponseEntity.ok(Collections.emptyList());
		}
		return ResponseEntity.ok(skillTypeRepository.findAllByGameType(byId.get()));
	}

	@PostMapping("/api/characterCreation/{gameId}")
	public ResponseEntity prepareCharacterCreation(@PathVariable Integer id, String characterName) {
		Optional<GameType> byId = gameTypeRepository.findById(id);
		if(!byId.isPresent()) {
			return ResponseEntity.notFound().build();
		}

		sessionStorage.prepareCharacterCreation(characterName, byId.get());
		return ResponseEntity.accepted().build();
	}

	@GetMapping(value = "/api/game_types/{gameId}/skillCreation", produces = MediaType.APPLICATION_JSON_UTF8_VALUE)
	public ResponseEntity<List<CreationDto>> getAllSkillCreations() {
		return ResponseEntity.ok(sessionStorage.getSkillCreationDtoList());
	}

	@GetMapping(value = "/api/game_types/{gameId}/baseValueCreation", produces = MediaType.APPLICATION_JSON_UTF8_VALUE)
	public ResponseEntity<List<CreationDto>> getAllBaseValueCreations() {
		return ResponseEntity.ok(sessionStorage.getBaseValueCreationDtoList());
	}

	@GetMapping(value = "/api/game_types/{gameId}/abilities", produces = MediaType.APPLICATION_JSON_UTF8_VALUE)
	public ResponseEntity<List<AttributeType>> getAllAttributes(@PathVariable Integer gameId) {
		Optional<GameType> byId = gameTypeRepository.findById(gameId);
		if (!byId.isPresent()) {
			return ResponseEntity.ok(Collections.emptyList());
		}
		return ResponseEntity.ok(attributeTypeRepository.findAllByGameType(byId.get()));
	}

	@GetMapping(value = "/api/game_types/{gameId}/baseValues", produces = MediaType.APPLICATION_JSON_UTF8_VALUE)
	public ResponseEntity<List<BaseValueType>> getAllBaseValues(@PathVariable Integer gameId) {
		Optional<GameType> byId = gameTypeRepository.findById(gameId);
		if (!byId.isPresent()) {
			return ResponseEntity.ok(Collections.emptyList());
		}
		return ResponseEntity.ok(baseValueRepository.findAllByGameType(byId.get()));
	}

	@GetMapping(value = "/api/game_types/{gameId}/details", produces = MediaType.APPLICATION_JSON_UTF8_VALUE)
	public ResponseEntity<List<DetailType>> getAllDetails(@PathVariable Integer gameId) {
		Optional<GameType> byId = gameTypeRepository.findById(gameId);
		if (!byId.isPresent()) {
			return ResponseEntity.ok(Collections.emptyList());
		}
		return ResponseEntity.ok(detailTypeRepository.findAllByGameType(byId.get()));
	}

	@PostMapping(value = "/api/game_types/{gameId}/skills", produces = MediaType.APPLICATION_JSON_UTF8_VALUE)
	public ResponseEntity newSkillType(String name, String description, String expression, boolean skillRequired, @PathVariable Integer gameId) {
		Optional<GameType> gameType = gameTypeRepository.findById(gameId);
		if (!gameType.isPresent()) {
			return new ResponseEntity(HttpStatus.NOT_FOUND);
		}

		if (skillTypeRepository.findByName(name).isPresent()) {
			return new ResponseEntity(HttpStatus.CONFLICT);
		}

		SkillType skillType = new SkillType();
		skillType.setRequired(skillRequired);
		skillType.setGameType(gameType.get());
		skillType.setName(name);
		skillType.setExpression(expression);
		skillType.setDescription(description);

		skillTypeRepository.save(skillType);
		List<SkillType> skillTypes = new ArrayList<>();
		skillTypeRepository.findAll().forEach(skillTypes::add);
		return ResponseEntity.accepted().body(new SkillTypeList(skillTypes));
	}

	@PostMapping(value = "/api/game_types/{gameId}/abilities", produces = MediaType.APPLICATION_JSON_UTF8_VALUE)
	public ResponseEntity newAttributeType(String name, String description, boolean abilityRequired, @PathVariable Integer gameId) {
		Optional<GameType> gameType = gameTypeRepository.findById(gameId);
		if (!gameType.isPresent()) {
			return new ResponseEntity(HttpStatus.NOT_FOUND);
		}

		AttributeType attributeType = new AttributeType();
		attributeType.setGameType(gameType.get());
		attributeType.setRequired(abilityRequired);
		attributeType.setName(name);
		attributeType.setDescription(description);

		attributeTypeRepository.save(attributeType);
		List<AttributeType> abilityTypes = new ArrayList<>();
		attributeTypeRepository.findAll().forEach(abilityTypes::add);
		return ResponseEntity.ok(abilityTypes);
	}

	@PostMapping(value = "/api/game_types/{gameId}/details", produces = MediaType.APPLICATION_JSON_UTF8_VALUE)
	public ResponseEntity newCharacterDetail(String name, String description, boolean detailRequired, @PathVariable Integer gameId) {
		Optional<GameType> gameType = gameTypeRepository.findById(gameId);
		if (!gameType.isPresent()) {
			return new ResponseEntity(HttpStatus.NOT_FOUND);
		}

		DetailType detailType = new DetailType(gameType.get(), name, description);
		detailType.setRequired(detailRequired);
		detailTypeRepository.save(detailType);
		return new ResponseEntity(HttpStatus.OK);
	}

	@PostMapping(value = "/api/game_types/{gameId}/baseValues", produces = MediaType.APPLICATION_JSON_UTF8_VALUE)
	public ResponseEntity newBaseValue(String name, String description, String expression, boolean baseValueRequired, @PathVariable Integer gameId) {
		Optional<GameType> gameType = gameTypeRepository.findById(gameId);
		if (!gameType.isPresent()) {
			return new ResponseEntity(HttpStatus.NOT_FOUND);
		}

		BaseValueType baseValueType = new BaseValueType();
		baseValueType.setGameType(gameType.get());
		baseValueType.setRequired(baseValueRequired);
		baseValueType.setName(name);
		baseValueType.setDescription(description);
		baseValueType.setExpression(expression);

		baseValueRepository.save(baseValueType);
		return new ResponseEntity(HttpStatus.OK);
	}

}

File: src/main/kotlin/com/github/thorben/webnd/application/controller/api/GameSpecificApi.kt
package com.github.thorben.webnd.application.controller.api

import com.github.thorben.webnd.application.SessionStorage
import com.github.thorben.webnd.application.controller.api.dto.SkillTypeList
import com.github.thorben.webnd.application.dto.CreationDto
import com.github.thorben.webnd.domain.character.*
import com.github.thorben.webnd.domain.games.GameType
import com.github.thorben.webnd.domain.games.GameTypeRepository
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RestController
import java.util.*
import java.util.function.Consumer

@RestController
class GameSpecificApi(
		private val skillTypeRepository: SkillTypeRepository,
		private val detailTypeRepository: DetailTypeRepository,
		private val attributeTypeRepository: AttributeTypeRepository,
		private val gameTypeRepository: GameTypeRepository,
		private val baseValueRepository: BaseValueTypeRepository,
		private val sessionStorage: SessionStorage
) {


	@GetMapping(value = ["/api/game_types"], produces = [MediaType.APPLICATION_JSON_UTF8_VALUE])
	fun allGameTypes(): ResponseEntity<List<GameType>> {
		val result: MutableList<GameType> = ArrayList()
		gameTypeRepository.findAll().forEach(Consumer { e: GameType -> result.add(e) })
		return ResponseEntity.ok(result)
	}

	@GetMapping(value = ["/api/game_types/{gameId}/skills"], produces = [MediaType.APPLICATION_JSON_UTF8_VALUE])
	fun getAllSkills(@PathVariable gameId: Int): ResponseEntity<List<SkillType>> {
		val byId = gameTypeRepository.findById(gameId)
		return if (!byId.isPresent) {
			ResponseEntity.ok(emptyList())
		} else {
			ResponseEntity.ok(skillTypeRepository.findAllByGameType(byId.get()))
		}
	}

	@PostMapping("/api/characterCreation/{gameId}")
	fun prepareCharacterCreation(@PathVariable id: Int, characterName: String): ResponseEntity<*> {
		val byId = gameTypeRepository.findById(id)
		if (!byId.isPresent) {
			return ResponseEntity.notFound().build<Any>()
		}
		sessionStorage.prepareCharacterCreation(characterName, byId.get())
		return ResponseEntity.accepted().build<Any>()
	}

	@get:GetMapping(value = ["/api/game_types/{gameId}/skillCreation"], produces = [MediaType.APPLICATION_JSON_UTF8_VALUE])
	val allSkillCreations: ResponseEntity<List<CreationDto>>
		get() = ResponseEntity.ok(sessionStorage.skillCreationDtoList)

	@get:GetMapping(value = ["/api/game_types/{gameId}/baseValueCreation"], produces = [MediaType.APPLICATION_JSON_UTF8_VALUE])
	val allBaseValueCreations: ResponseEntity<List<CreationDto>>
		get() = ResponseEntity.ok(sessionStorage.baseValueCreationDtoList)

	@GetMapping(value = ["/api/game_types/{gameId}/abilities"], produces = [MediaType.APPLICATION_JSON_UTF8_VALUE])
	fun getAllAttributes(@PathVariable gameId: Int): ResponseEntity<List<AttributeType?>?> {
		val byId = gameTypeRepository.findById(gameId)
		return if (!byId.isPresent) {
			ResponseEntity.ok(emptyList())
		} else ResponseEntity.ok(attributeTypeRepository.findAllByGameType(byId.get()))
	}

	@GetMapping(value = ["/api/game_types/{gameId}/baseValues"], produces = [MediaType.APPLICATION_JSON_UTF8_VALUE])
	fun getAllBaseValues(@PathVariable gameId: Int): ResponseEntity<List<BaseValueType>> {
		val byId = gameTypeRepository.findById(gameId)
		return if (!byId.isPresent) {
			ResponseEntity.ok(emptyList())
		} else {
			ResponseEntity.ok(baseValueRepository.findAllByGameType(byId.get()))
		}
	}

	@GetMapping(value = ["/api/game_types/{gameId}/details"], produces = [MediaType.APPLICATION_JSON_UTF8_VALUE])
	fun getAllDetails(@PathVariable gameId: Int): ResponseEntity<List<DetailType>> {
		val byId = gameTypeRepository.findById(gameId)
		return if (!byId.isPresent) {
			ResponseEntity.ok(emptyList())
		} else {ResponseEntity.ok(detailTypeRepository.findAllByGameType(byId.get()))}
	}

	@PostMapping(value = ["/api/game_types/{gameId}/skills"], produces = [MediaType.APPLICATION_JSON_UTF8_VALUE])
	fun newSkillType(name: String, description: String, expression: String, skillRequired: Boolean, @PathVariable gameId: Int): ResponseEntity<*> {
		val gameType = gameTypeRepository.findById(gameId)
		if (!gameType.isPresent) {
			return ResponseEntity<Any?>(HttpStatus.NOT_FOUND)
		}
		if (skillTypeRepository.findByName(name).isPresent) {
			return ResponseEntity<Any?>(HttpStatus.CONFLICT)
		}
		val skillType = SkillType()
		skillType.isRequired = skillRequired
		skillType.gameType = gameType.get()
		skillType.name = name
		skillType.expression = expression
		skillType.description = description
		skillTypeRepository.save(skillType)
		val skillTypes: MutableList<SkillType> = ArrayList()
		skillTypeRepository.findAll().forEach(Consumer { e: SkillType -> skillTypes.add(e) })
		return ResponseEntity.accepted().body(SkillTypeList(skillTypes))
	}

	@PostMapping(value = ["/api/game_types/{gameId}/abilities"], produces = [MediaType.APPLICATION_JSON_UTF8_VALUE])
	fun newAttributeType(name: String, description: String, abilityRequired: Boolean, @PathVariable gameId: Int): ResponseEntity<*> {
		val gameType = gameTypeRepository.findById(gameId)
		if (!gameType.isPresent) {
			return ResponseEntity<Any?>(HttpStatus.NOT_FOUND)
		}
		val attributeType = AttributeType()
		attributeType.gameType = gameType.get()
		attributeType.isRequired = abilityRequired
		attributeType.name = name
		attributeType.description = description
		attributeTypeRepository.save(attributeType)
		val abilityTypes: MutableList<AttributeType> = ArrayList()
		attributeTypeRepository.findAll().forEach { abilityTypes.add(it) }
		return ResponseEntity.ok<List<AttributeType>>(abilityTypes)
	}

	@PostMapping(value = ["/api/game_types/{gameId}/details"], produces = [MediaType.APPLICATION_JSON_UTF8_VALUE])
	fun newCharacterDetail(name: String, description: String, detailRequired: Boolean, @PathVariable gameId: Int): ResponseEntity<*> {
		val gameType = gameTypeRepository.findById(gameId)
		if (!gameType.isPresent) {
			return ResponseEntity<Any?>(HttpStatus.NOT_FOUND)
		}
		val detailType = DetailType(gameType.get(), name, description)
		detailType.isRequired = detailRequired
		detailTypeRepository.save(detailType)
		return ResponseEntity<Any?>(HttpStatus.OK)
	}

	@PostMapping(value = ["/api/game_types/{gameId}/baseValues"], produces = [MediaType.APPLICATION_JSON_UTF8_VALUE])
	fun newBaseValue(name: String, description: String, expression: String, baseValueRequired: Boolean, @PathVariable gameId: Int): ResponseEntity<*> {
		val gameType = gameTypeRepository.findById(gameId)
		if (!gameType.isPresent) {
			return ResponseEntity<Any?>(HttpStatus.NOT_FOUND)
		}
		val baseValueType = BaseValueType()
		baseValueType.gameType = gameType.get()
		baseValueType.isRequired = baseValueRequired
		baseValueType.name = name
		baseValueType.description = description
		baseValueType.expression = expression
		baseValueRepository.save(baseValueType)
		return ResponseEntity<Any?>(HttpStatus.OK)
	}
}